jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3411,9 +3413,54 @@ private static MethodHandle validateArgumentsOfWhileLoop(MethodHandle predHandle
 		Class<?> bodyReturnType = bodyType.returnType;
 		
 		/* The signature of the body handle must be either (V, A...)V or (A...)void */
-		if ((void.class != bodyReturnType) && (bodyReturnType != bodyType.arguments[0])) {
-			/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
-			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), bodyType.arguments[0].getName())); //$NON-NLS-1$
+		int bodyVlength = 0;
+		if (void.class != bodyReturnType) {
+			if (hasNoArgs(bodyType) 
+					|| (bodyReturnType != bodyType.arguments[0])) {
+				String argList = "void";
+				if (! hasNoArgs(bodyType)) {
+					argList = bodyType.arguments[0].getName();
+				}
+				
+				/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), argList)); //$NON-NLS-1$
+			}
+			bodyVlength = 1;
+		}
+		
+		/* Verify that handles are effectively identical and constrained by body. All we
+		 * need to do here is verify that no parameters have longer argument lists than
+		 * body. The generic loop will compare arguments to ensure all clauses are
+		 * effectively identical.
+		 */
+		int bodyParamTypesLength = bodyHandle.type.arguments.length;
+		int bodyAlength = bodyParamTypesLength - bodyVlength;
+		
+		/* predHandle parameters are empty or (V, A..). Make sure it is compatible with
+		 * the effectively identical model that compares A. An empty list can be skipped
+		 * as it is always effectively identical.
+		 */
+		MethodType predParamType = predHandle.type;
+		if (! hasNoArgs(predParamType)) {
+			Class<?>[] predParamTypes = predHandle.type.arguments;
+			
+			if (void.class != bodyReturnType) {
+				if (bodyReturnType != predParamTypes[0]) {		
+					/*[MSG "K0654", "Pred parameter list must be either empty or of the form (V A*)"]*/		
+					throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0654")); //$NON-NLS-1$		
+				}
+			}
+			
+			int externalPredParamTypesLength =  predParamType.parameterCount() - bodyVlength;
+			if (externalPredParamTypesLength > bodyAlength) {
+				/*[MSG "K065K", "Loop parameter {0} must be effectively identical to body"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065K", "pred"));  //$NON-NLS-2$
+			}
+		}
+		
```

Comment:
Also, do you check that the predicate returnType must always be a `boolean` ?

Reply:
That will also be checked in the generic loop. `loop()` has a rule "Every non-omitted pred function must have a boolean return type.". The pred parameters for `doWhileLoop` and `whileLoop` are both passed into `loop` as a pred (third slot in a clause).

Reply:
This is also fixed.